### PR TITLE
orca: add application utilization and range checking

### PR DESCRIPTION
**Pull request by dfawley**
Implement the new field described [here](https://github.com/grpc/proposal/pull/374), and add missing range checks and tests.

RELEASE NOTES:
* orca: add support for application_utilization field and missing range checks on several metrics setters

### CODE REVIEW

File: orca/server_metrics.go [hunk 1]
```diff
@@ -194,36 +221,70 @@ func (s *serverMetricsRecorder) SetMemoryUtilization(val float64) {
 // DeleteMemoryUtilization deletes the relevant server metric to prevent it
 // from being sent.
```
Comment (zasweq) [inline orca/server_metrics.go:222]: Nit on these docstrings and also the interface docstrings. Most of them define "the relevant server metric", but I feel like it should just be as specific as "for the memory utilization metric" as on line 211 for all docstrings because it maps 1:1 with the name.

File: orca/server_metrics.go [hunk 2]
```diff
@@ -181,11 +203,16 @@ func (s *serverMetricsRecorder) SetCPUUtilization(val float64) {
...
-	s.SetCPUUtilization(-1)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.state.CPUUtilization = -1
 }
 
 // SetMemoryUtilization records a measurement for the memory utilization metric.
 func (s *serverMetricsRecorder) SetMemoryUtilization(val float64) {
+	if val < 0 || val > 1 {
+		return
```
Comment (zasweq) [inline orca/server_metrics.go:214]: Here and all other places with a range validation on this serverMetricsRecorder: do we want to plumb a logger in here and log in a non spammy way that we're ignoring an out of range metric, similar to how xDS Client gives info if nack, or say CDS encounters an error in it's serial handling of Client Conn State that returns no error, it logs the reason why the Client Conn State wasn't processed?

File: orca/server_metrics.go [hunk 3]
```diff
@@ -63,21 +67,24 @@ func (sm *ServerMetrics) toLoadReportProto() *v3orcapb.OrcaLoadReport {
 
 // merge merges o into sm, overwriting any values present in both.
 func (sm *ServerMetrics) merge(o *ServerMetrics) {
+	mergeMap(sm.Utilization, o.Utilization)
+	mergeMap(sm.RequestCost, o.RequestCost)
+	mergeMap(sm.NamedMetrics, o.NamedMetrics)
```
Comment (zasweq) [inline orca/server_metrics.go:72]: What happens if one of the values of this map is also -1 like below? Do we still want to overwrite the value for the key in the map? Should we ignore as below?

File: orca/server_metrics.go [hunk 4]
```diff
@@ -63,21 +67,24 @@ func (sm *ServerMetrics) toLoadReportProto() *v3orcapb.OrcaLoadReport {
 
 // merge merges o into sm, overwriting any values present in both.
```
Comment (zasweq) [inline orca/server_metrics.go:68]: This isn't correct? You overwrite any values for keys in map unconditionally, but only overwrite sm's memory if it's set. I guess if an explicit -1 means unset 1:1 this is right. I just don't know all the invariants of this system/how this data is plumbed around.

File: orca/producer_test.go [hunk 5]
```diff
@@ -444,8 +444,8 @@ func (s) TestProducerMultipleListeners(t *testing.T) {
 	// Define a load report to send and expect the client to see.
 	loadReportWant := &v3orcapb.OrcaLoadReport{
 		CpuUtilization: 10,
-		MemUtilization: 100,
-		Utilization:    map[string]float64{"bob": 555},
+		MemUtilization: 0.1,
```
Comment (zasweq) [inline orca/producer_test.go:447]: Now that you had to move these floats to be within 0 and 1, can you please add a test for the case it isn't within range and also set to -1 wrt overwriting behavior in merge(). This will also show what happens at this producer layer if these cases happen.

File: orca/server_metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+	// Negatives for all these fields should be ignored.
+	smr.SetCPUUtilization(-2)
+	smr.SetMemoryUtilization(-3)
+	smr.SetApplicationUtilization(-4)
+	smr.SetQPS(-0.1)
+	smr.SetEPS(-0.6)
+	smr.SetNamedUtilization("x", -2)
+
+	// Memory and named utilizations are max of 1.
```
Comment (zasweq) [inline orca/server_metrics_test.go:109]: So this should be ignored.

File: orca/server_metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		Utilization:    map[string]float64{"x": 0.6},
+		NamedMetrics:   map[string]float64{},
+		RequestCost:    map[string]float64{},
+	}
+
+	got := smr.ServerMetrics()
+	if d := cmp.Diff(got, want); d != "" {
+		t.Fatalf("unexpected server metrics: -got +want: %v", d)
+	}
+}
```
Comment (zasweq) [inline orca/server_metrics_test.go:128]: After thinking about my comment on merge() wrt overwriting at each layer (maps and top level metrics), can you also add tests for that.

File: orca/service_test.go [hunk 8]
```diff
@@ -60,16 +60,18 @@ func (t *testServiceImpl) UnaryCall(context.Context, *testpb.SimpleRequest) (*te
 	t.requests++
 	t.mu.Unlock()
 
-	t.smr.SetNamedUtilization(requestsMetricKey, float64(t.requests))
+	t.smr.SetNamedUtilization(requestsMetricKey, float64(t.requests)*0.01)
```
Comment (zasweq) [inline orca/service_test.go:63]: Like my comment above, what happens in this layer if these metrics aren't within range? Can you please add a test to show what behavior is going to be.

**Review by zasweq — CHANGES_REQUESTED** [review]

File: orca/server_metrics.go (hunk 3 above)
Comment (dfawley) [reply orca/server_metrics.go:72]: For maps, `-1` has no special meaning.  Deleting from the map is how you delete the value.  -1 is unique to the fields that are scalar values.

File: orca/server_metrics.go (hunk 4 above)
Comment (dfawley) [reply orca/server_metrics.go:68]: > You overwrite any values for keys in map unconditionally, but only overwrite sm's memory if it's set.

No?  I range over `o.<map field>` and add all those key/value pairs into `sm.<map field>`.  It's a merge, not a replacement, so the idea is to keep anything in `sm` that's not in `o`.

File: orca/producer_test.go (hunk 5 above)
Comment (dfawley) [reply orca/producer_test.go:447]: I have range checks in the new tests.

File: orca/server_metrics_test.go (hunk 6 above)
Comment (dfawley) [reply orca/server_metrics_test.go:109]: Clarified.

File: orca/server_metrics_test.go (hunk 7 above)
Comment (dfawley) [reply orca/server_metrics_test.go:128]: Added unit tests for merge..  It is already covered implicitly by the call metrics tests that use it, though.

File: orca/service_test.go (hunk 8 above)
Comment (dfawley) [reply orca/service_test.go:63]: This is covered in the unit tests.  It's just calls into the same object so if that works right then it must work right here, too.

File: orca/server_metrics.go (hunk 3 above)
Comment (zasweq) [reply orca/server_metrics.go:72]: Ah ok.

File: orca/producer_test.go (hunk 5 above)
Comment (zasweq) [reply orca/producer_test.go:447]: I know, but what happens at this layer? Just ignored? I want it tested at this layer too, or are you saying specific behaviors are fine to encapsulate to it's granular components units.

File: orca/service_test.go (hunk 8 above)
Comment (zasweq) [reply orca/service_test.go:63]: Fine.

Comment (zasweq) [reply orca/service_test.go:63]: This relates to discussions we've had and I've had with Eric about specific functionality scoped to the granular component for testing, and higher layers just do sanity to avoid excessive higher level tests. I'm actually running into this problem with my od change, since ParseConfig gets invoked at every layer of unit test, whether it's cds, cluster resolver, and obv. on the od ParseConfig test itself.

Comment (dfawley) [reply orca/service_test.go:63]: ...Which is why I was saying to test the proto->JSON conversion as a unit test so you can get all the corner cases in there and not worry about it when testing the higher level logic.

File: orca/server_metrics.go [hunk 9]
```diff
@@ -193,6 +195,9 @@ func copyMap(m map[string]float64) map[string]float64 {
 // SetCPUUtilization records a measurement for the CPU utilization metric.
 func (s *serverMetricsRecorder) SetCPUUtilization(val float64) {
 	if val < 0 {
+		if logger.V(2) {
+			logger.Infof("Ignoring CPU Utilization value out of range: %v", val)
```
Comment (zasweq) [inline orca/server_metrics.go:199]: Log valid range here and elsewhere?

File: orca/server_metrics_test.go [hunk 10]
```diff
@@ -126,3 +134,42 @@ func (s) TestServerMetrics_Setters_Range(t *testing.T) {
...
+		QPS:            -1,
+		EPS:            0,
+		Utilization:    map[string]float64{"x": 0.6},
+		NamedMetrics:   map[string]float64{"y": 0.2},
+		RequestCost:    map[string]float64{"a": 0.1},
+	}
+
+	sm2 := &ServerMetrics{
+		CPUUtilization: -1,
+		AppUtilization: 0,
```
Comment (zasweq) [inline orca/server_metrics_test.go:152]: Unfortunate that MemUtilization's zero value = set valid metric which overwrites, but can't do anything about that.

**Review by zasweq — APPROVED** [review]
LGTM.

File: orca/server_metrics.go (hunk 9 above)
Comment (dfawley) [reply orca/server_metrics.go:199]: Ehh, they'll figure it out :)

File: orca/server_metrics_test.go (hunk 10 above)
Comment (dfawley) [reply orca/server_metrics_test.go:152]: Yeah the only option would be to use pointers and I don't really want to do that either.  We basically control everything around this struct and the values of the fields, though, so it's pretty safe.

File: orca/server_metrics.go (hunk 9 above)
Comment (zasweq) [reply orca/server_metrics.go:199]: "[0, infy)" this is very few characters to add to logs lol
